codegen-client/src/test/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/waiters/RustWaiterMatcherGeneratorTest.kt

```diff
@@ -0,0 +1,349 @@
+import software.amazon.smithy.rust.codegen.client.testutil.clientIntegrationTest
+import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.rustlang.Writable
+import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.writable
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.preludeScope
+import software.amazon.smithy.rust.codegen.core.smithy.RustCrate
+import software.amazon.smithy.rust.codegen.core.testutil.asSmithyModel
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.inputShape
+import software.amazon.smithy.rust.codegen.core.util.lookup
+import software.amazon.smithy.rust.codegen.core.util.outputShape
+import software.amazon.smithy.waiters.Matcher.SuccessMember
+
+private typealias Scope = Array<Pair<String, Any>>
+
+class RustWaiterMatcherGeneratorTest {
+    class TestCase(
+        codegenContext: ClientCodegenContext,
+        private val rustCrate: RustCrate,
+        matcherJson: String,
+    ) {
+        val operationShape = codegenContext.model.lookup<OperationShape>("test#TestOperation")
+        val inputShape = operationShape.inputShape(codegenContext.model)
+        val outputShape = operationShape.outputShape(codegenContext.model)
+        val errorShape = codegenContext.model.lookup<StructureShape>("test#SomeError")
+        val inputSymbol = codegenContext.symbolProvider.toSymbol(inputShape)
+        val outputSymbol = codegenContext.symbolProvider.toSymbol(outputShape)
+        val errorSymbol = codegenContext.symbolProvider.toSymbol(errorShape)
+
+        val matcher = SuccessMember.fromNode(Node.parse(matcherJson))
+        val matcherFn =
+            RustWaiterMatcherGenerator(codegenContext, "TestOperation", inputShape, outputShape)
+                .generate(errorSymbol, matcher)
+
+        val scope =
+            arrayOf(
+                *preludeScope,
+                "Input" to inputSymbol,
+                "Output" to outputSymbol,
+                "Error" to errorSymbol,
+                "ErrorMetadata" to RuntimeType.errorMetadata(codegenContext.runtimeConfig),
+                "matcher_fn" to matcherFn,
+            )
+
+        fun renderTest(
+            name: String,
+            writeTest: TestCase.() -> Writable,
+        ) {
+            rustCrate.lib {
+                rustTemplate(
+                    """
+                    /// Make the unit test public and document it so that compiler
+                    /// doesn't complain about dead code.
+                    pub fn ${name}_test_case() {
+                        #{test}
+                    }
+                    ##[cfg(test)]
+                    ##[test]
+                    fn $name() {
+                        ${name}_test_case();
```

Comment:
**question**: Why not just generate the test body `#{test}` here and avoid the need to make it public/documented?

Reply:
The compiler is unhappy about the `pub(crate)` function under test. Since its only used in a unit test, it considers it dead code, and since we have warnings as errors, this causes the test to fail.

Reply:
Ahh I see